compat/unistd.cc

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 1996-2025 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#include "squid.h"
+
+#if _SQUID_WINDOWS_ || _SQUID_MINGW_
+
+#include "compat/socket.h"
+#include "compat/unistd.h"
+#include "compat/wserrno.h"
+
+// 2025 MinGW and pre-2022 MSVC do not define these
+#ifndef _S_IREAD
+#define _S_IREAD 0x0100
+#endif
+
+#ifndef _S_IWRITE
+#define _S_IWRITE 0x0080
+#endif
+
+/// returns true if handle is a valid socket. Preserves errno.
+static bool
+isSocket(intptr_t handle)
+{
+    if (!isValidSocketHandle(handle)) {
+        // isValidSocketHandle does not touch errno
+        return false;
+    }
+
+    int value = 0;
+    int valueSize = sizeof(value);
+    // use the Windows API directly
+    auto saved_errno = errno;
+    if (getsockopt(handle, SOL_SOCKET, SO_TYPE, reinterpret_cast<char *>(&value), &valueSize) == 0) {
+        errno = saved_errno;
+        return true;
+    }
+
+    errno = saved_errno;
+    return false;
```

Comment:
Please use `const` and camelCase. I recommend the following simple `if`-less code structure:

```suggestion
    const auto savedErrno = errno;
    // use Windows API directly
    const auto result = (getsockopt(handle, SOL_SOCKET, SO_TYPE, reinterpret_cast<char *>(&value), &valueSize) == 0);
    errno = savedErrno;
    return result;
```

You can even move isValidSocketHandle() check into `result` calculation, avoiding documenting implicit/current isValidSocketHandle() guarantees in its caller, but I do not insist on that.